src/execution/window_executor.cpp

```diff
@@ -7,6 +7,75 @@
 
 namespace duckdb {
 
+//===--------------------------------------------------------------------===//
+// WindowDataChunk
+//===--------------------------------------------------------------------===//
+bool WindowDataChunk::IsSimple(const Vector &v) {
+	switch (v.GetType().InternalType()) {
+	case PhysicalType::BOOL:
+	case PhysicalType::UINT8:
+	case PhysicalType::INT8:
+	case PhysicalType::UINT16:
+	case PhysicalType::INT16:
+	case PhysicalType::UINT32:
+	case PhysicalType::INT32:
+	case PhysicalType::UINT64:
+	case PhysicalType::INT64:
+	case PhysicalType::FLOAT:
+	case PhysicalType::DOUBLE:
+	case PhysicalType::INTERVAL:
+	case PhysicalType::LIST:
+	case PhysicalType::UINT128:
+	case PhysicalType::INT128:
+		return true;
+	case PhysicalType::STRUCT:
+	case PhysicalType::ARRAY:
+	case PhysicalType::VARCHAR:
+	case PhysicalType::BIT:
+		return false;
+	default:
+		break;
+	}
+
+	throw InternalException("Unsupported type for WindowDataChunk");
+}
+
+WindowDataChunk::WindowDataChunk(DataChunk &chunk) : chunk(chunk) {
+}
+
+void WindowDataChunk::Initialize(Allocator &allocator, const vector<LogicalType> &types, idx_t capacity) {
+	vector<mutex> new_locks(types.size());
+	locks.swap(new_locks);
+	chunk.Initialize(allocator, types, capacity);
+	chunk.SetCardinality(capacity);
+
+	is_simple.clear();
+	for (const auto &v : chunk.data) {
+		is_simple.push_back(IsSimple(v));
+	}
+}
+
+void WindowDataChunk::Copy(DataChunk &input, idx_t begin) {
+	const auto source_count = input.size();
+	const idx_t end = begin + source_count;
+	const idx_t count = chunk.size();
+	D_ASSERT(end <= count);
+	// Can we overwrite the validity mask in parallel?
+	bool aligned = IsMaskAligned(begin, end, count);
+	for (column_t i = 0; i < chunk.data.size(); ++i) {
+		auto &src = input.data[i];
+		auto &dst = chunk.data[i];
+		UnifiedVectorFormat sdata;
+		src.ToUnifiedFormat(count, sdata);
+		if (is_simple[i] && aligned && sdata.validity.AllValid()) {
+			VectorOperations::Copy(src, dst, source_count, 0, begin);
+		} else {
+			lock_guard<mutex> column_guard(locks[i]);
```

Comment:
Perhaps it makes sense to switch away from `DataChunk` here so that we can avoid these locks?

Looking at the operations that are done on `WindowDataChunk` - I think it would be rather straightforward to replace the `DataChunk` with a raw array of the underlying data type, together with a `bool[]` array for the nullmask.

Nested types are the only complexity here - but now that we have the sort keys that is not too complex either - we can serialize the sort keys on ingestion, and deserialize when we want to access the value (i.e. in `GetCell`). The [`CreateSortKeyHelpers`](https://github.com/duckdb/duckdb/blob/fd883da87821a3e4a60169df7cba49f29d7a252e/src/include/duckdb/core_functions/create_sort_key.hpp#L48) can make this easy. Then all we need to handle are primitive types (`int32_t`, `int64_t`, etc) and `string_t`.

Reply:
I think the issue would be that some of the accelerators (e.g., segment trees) _really_ want a `DataChunk` at the bottom level. My thinking here was that `Vector` is already mostly a pointer to an array of type and the only messy thing(!) is the validity mask. To improve the mask situation I was thinking of

* Aligning the scanned blocks somehow (e.g., skipping ragged beginnings, or maybe even @lnkuiper 's ICDE structures would naturally be this way?)
* Checking to see if there is already a mask, in which case alignment alone is sufficient.

This is just our first step towards having larger in-memory data structures like Hyper/Umbra do. Where it gets really interesting is the _next_ project where we replace them with some kind of pageable columnar data structure to further reduce (or eliminate?!) memory pressure.

Reply:
We can turn a raw array back into a `Vector` using [this constructor](https://github.com/duckdb/duckdb/blob/fd883da87821a3e4a60169df7cba49f29d7a252e/src/include/duckdb/common/types/vector.hpp#L107) - although the validity mask would have to be translated separately.

> This is just our first step towards having larger in-memory data structures like Hyper/Umbra do. Where it gets really interesting is the next project where we replace them with some kind of pageable columnar data structure to further reduce (or eliminate?!) memory pressure.

We have this in the form of the `ColumnDataCollection` already of course, the main reason not to use it here (IIRC) is the lack of fast random access. Are you thinking about designing another structure that does enable fast random access in some capacity, or perhaps extending the `ColumnDataCollection` with better support for it?

Reply:
Yes, but I haven't looked at it much yet. Partly because I think we have several such data structures now and are trying to reduce the number first?

Reply:
(Oh and by "next project" I meant later this year, not later this month!)

Reply:
The `ChunkCollection` has been removed - so for columnar we only have the `ColumnDataCollection` currently. For rows we have the `RowDataCollection` and the `TupleDataCollection`, with the `RowDataCollection` being on the chopping block.

Reply:
`RowDataCollection` is one of my pain points right now...